File: src/lib/narration.ts
import { Directory, File, Paths } from 'expo-file-system';

import type { SegmentNarration, Story, StoryNarrationSummary } from '@/types/story';

/**
 * Narration resolution — the heart of the hybrid recorded/cloned model.
 *
 * The unit of audio is the SEGMENT: a run of text spoken by one voice
 * (the narrator or a character). Each segment slot is filled by, in order:
 *   1. a real recording made by the profile owner  → kind 'recorded'
 *   2. cloned TTS in the matching voice (future)   → kind 'cloned'
 *   3. nothing yet                                 → kind 'none'
 *
 * Because every character segment is recorded against known text and a known
 * character, recordings double as labeled training samples for future
 * character voice models (see lib/cast.ts).
 *
 * Recordings live on-device under:
 *   <documents>/recordings/<profileId>/<storyId>/p<page>-s<segment>.m4a
 */

function storyDir(profileId: string, storyId: string): Directory {
  return new Directory(Paths.document, 'recordings', profileId, storyId);
}

export function segmentRecordingFile(
  profileId: string,
  storyId: string,
  pageIndex: number,
  segIndex: number
): File {
  return new File(storyDir(profileId, storyId), `p${pageIndex}-s${segIndex}.m4a`);
}

export async function saveSegmentRecording(
  profileId: string,
  storyId: string,
  pageIndex: number,
  segIndex: number,
  sourceUri: string
): Promise<string> {
  const dir = storyDir(profileId, storyId);
  dir.create({ intermediates: true, idempotent: true });

  const dest = segmentRecordingFile(profileId, storyId, pageIndex, segIndex);
  if (dest.exists) {
    dest.delete();
  }
  await new File(sourceUri).copy(dest);
  return dest.uri;
}

export function deleteSegmentRecording(
  profileId: string,
  storyId: string,
  pageIndex: number,
  segIndex: number
): void {
  const file = segmentRecordingFile(profileId, storyId, pageIndex, segIndex);
  if (file.exists) {
    file.delete();
  }
}

export function getSegmentNarration(
  profileId: string | null,
  storyId: string,
  pageIndex: number,
  segIndex: number
): SegmentNarration {
  if (!profileId) {
    return { kind: 'none', uri: null };
  }
  const recording = segmentRecordingFile(profileId, storyId, pageIndex, segIndex);
  if (recording.exists) {
    return { kind: 'recorded', uri: recording.uri };
  }
  // Future: fall back to cloned TTS — the narrator voice model for narrator
  // segments, or the matching character voice model from the profile's cast —
  // returning { kind: 'cloned', uri } so the UI can label the source.
  return { kind: 'none', uri: null };
}

/** Per page, per segment. */
export function getStoryNarration(profileId: string | null, story: Story): SegmentNarration[][] {
  return story.pages.map((page, pageIndex) =>
    page.segments.map((_, segIndex) => getSegmentNarration(profileId, story.id, pageIndex, segIndex))
  );
}

export function summarizeStory(profileId: string | null, story: Story): StoryNarrationSummary {
  const narrations = getStoryNarration(profileId, story).flat();
  const recorded = narrations.filter((n) => n.kind === 'recorded').length;
  const total = narrations.length;
  return {
    recorded,
    total,
    status: recorded === 0 ? 'none' : recorded === total ? 'complete' : 'partial',
  };
}

File: app/src/main/java/com/robbstacy/bedtimecast/data/Narration.kt
package com.robbstacy.bedtimecast.data

import android.content.Context
import java.io.File

/**
 * Narration storage — the heart of the hybrid recorded/cloned model.
 *
 * The unit of audio is the SEGMENT: a run of text spoken by one voice
 * (the narrator or a character). Recordings live on-device under:
 *   filesDir/recordings/<profileId>/<storyId>/p<page>-s<segment>.m4a
 *
 * Because every character segment is recorded against known text and a known
 * character, recordings double as labeled training samples for future
 * character voice models (see [cast]).
 */
object Narration {

    fun segmentFile(
        context: Context,
        profileId: String,
        storyId: String,
        page: Int,
        segment: Int,
    ): File {
        val dir = File(File(File(context.filesDir, "recordings"), profileId), storyId)
        return File(dir, "p$page-s$segment.m4a")
    }

    fun deleteSegment(context: Context, profileId: String, storyId: String, page: Int, segment: Int) {
        segmentFile(context, profileId, storyId, page, segment).delete()
    }

    /** Cloned (AI-generated) narration cache, mirroring the recordings layout. */
    fun clonedFile(
        context: Context,
        profileId: String,
        storyId: String,
        page: Int,
        segment: Int,
    ): File {
        val dir = File(File(File(context.filesDir, "cloned"), profileId), storyId)
        return File(dir, "p$page-s$segment.mp3")
    }

    enum class Kind { RECORDED, CLONED, NONE }

    data class SegmentNarration(val kind: Kind, val file: File?)

    /** Real recording first, then cached cloned audio, then nothing. */
    fun segmentNarration(
        context: Context,
        profileId: String?,
        storyId: String,
        page: Int,
        segment: Int,
    ): SegmentNarration {
        if (profileId == null) return SegmentNarration(Kind.NONE, null)
        val recorded = segmentFile(context, profileId, storyId, page, segment)
        if (recorded.exists()) return SegmentNarration(Kind.RECORDED, recorded)
        val cloned = clonedFile(context, profileId, storyId, page, segment)
        if (cloned.exists()) return SegmentNarration(Kind.CLONED, cloned)
        return SegmentNarration(Kind.NONE, null)
    }

    fun narrationMatrix(context: Context, profileId: String?, story: Story): List<List<SegmentNarration>> =
        story.pages.mapIndexed { pageIndex, page ->
            page.segments.indices.map { segIndex ->
                segmentNarration(context, profileId, story.id, pageIndex, segIndex)
            }
        }

    data class SampleStats(val lines: Int, val words: Int)

    /** Recorded narrator material across the whole library for one profile. */
    fun narratorStats(context: Context, profileId: String, stories: List<Story>): SampleStats {
        var lines = 0
        var words = 0
        for (story in stories) {
            story.pages.forEachIndexed { pageIndex, page ->
                page.segments.forEachIndexed { segIndex, segment ->
                    if (segment.speaker != "narrator") return@forEachIndexed
                    if (segmentFile(context, profileId, story.id, pageIndex, segIndex).exists()) {
                        lines++
                        words += segment.text.split(Regex("\\s+")).count { it.isNotBlank() }
                    }
                }
            }
        }
        return SampleStats(lines, words)
    }

    fun narratorFiles(context: Context, profileId: String, stories: List<Story>): List<File> {
        val files = mutableListOf<File>()
        for (story in stories) {
            story.pages.forEachIndexed { pageIndex, page ->
                page.segments.forEachIndexed { segIndex, segment ->
                    if (segment.speaker != "narrator") return@forEachIndexed
                    val file = segmentFile(context, profileId, story.id, pageIndex, segIndex)
                    if (file.exists()) files.add(file)
                }
            }
        }
        return files
    }

    fun characterFiles(
        context: Context,
        profileId: String,
        story: Story,
        characterId: String,
    ): List<File> {
        val files = mutableListOf<File>()
        story.pages.forEachIndexed { pageIndex, page ->
            page.segments.forEachIndexed { segIndex, segment ->
                if (segment.speaker != characterId) return@forEachIndexed
                val file = segmentFile(context, profileId, story.id, pageIndex, segIndex)
                if (file.exists()) files.add(file)
            }
        }
        return files
    }

    /** Per page, per segment: is a recording present? */
    fun recordedMatrix(context: Context, profileId: String?, story: Story): List<List<Boolean>> =
        story.pages.mapIndexed { pageIndex, page ->
            page.segments.indices.map { segIndex ->
                profileId != null &&
                    segmentFile(context, profileId, story.id, pageIndex, segIndex).exists()
            }
        }

    data class Summary(val recorded: Int, val total: Int) {
        val complete: Boolean get() = total > 0 && recorded == total
        val none: Boolean get() = recorded == 0
    }

    fun summarize(context: Context, profileId: String?, story: Story): Summary {
        val flat = recordedMatrix(context, profileId, story).flatten()
        return Summary(recorded = flat.count { it }, total = flat.size)
    }

    /** All characters the profile could perform, with recording coverage. */
    fun cast(context: Context, profileId: String?, stories: List<Story>): List<CastMember> {
        if (profileId == null) return emptyList()
        val members = mutableListOf<CastMember>()
        for (story in stories) {
            for (character in story.characters) {
                var totalLines = 0
                var recordedLines = 0
                var recordedWords = 0
                story.pages.forEachIndexed { pageIndex, page ->
                    page.segments.forEachIndexed { segIndex, segment ->
                        if (segment.speaker != character.id) return@forEachIndexed
                        totalLines++
                        if (segmentFile(context, profileId, story.id, pageIndex, segIndex).exists()) {
                            recordedLines++
                            recordedWords += segment.text.split(Regex("\\s+")).count { it.isNotBlank() }
                        }
                    }
                }
                if (totalLines > 0) {
                    members.add(CastMember(story, character, totalLines, recordedLines, recordedWords))
                }
            }
        }
        return members
    }
}
